- Part 1: Data Exploration with Pandas
Load the provided dataset into a pandas dataframe
Use pandas to explore the dataset. For example, you might start by using the head() and info() methods to get an overview of the data.
Look for missing values in the dataset. You can use the isnull() method to identify missing values.
Use pandas to calculate some summary statistics for the dataset. For example, you might use the describe() method to get summary statistics for the numerical columns in the dataset.
- Part 2: Data Exploration with ydata-Profiling
Use ydata-profiling to generate a report of the provided dataset.
Look for missing values in the dataset. You can use the report generated by ydata-profiling to identify missing values.
Look for correlations between different columns in the dataset. You can use the report generated by ydata-profiling to identify correlations between different columns.
Identify any outliers or unusual values in the dataset. You can use the report generated by ydata-profiling to identify any outliers or unusual values.
Summary
At the end of this exercise, write a summary of your findings. Did you find any interesting patterns or correlations in the data? Were there any issues or challenges you encountered while exploring the dataset? Use your summary to reflect on your experience using pandas and ydata-profiling to explore and understand a new dataset.

In [1]:
import pandas as pd
from ydata_profiling import ProfileReport

In [2]:
df_airtunisia = pd.read_csv('Tunisair_flights_dataset.csv')
df_airtunisia.head()

Filght_date Flight_ID Departure point Arrival point  \
0  2016-01-03  TU 0712              CMN           TUN   
1  2016-01-13  TU 0757              MXP           TUN   
2  2016-01-16  TU 0214              TUN           IST   
3  2016-01-17  TU 0480              DJE           NTE   
4  2016-01-17  TU 0338              TUN           ALG   

  Scheduled_departure_time Scheduled_arrival_time STATUS Aircraft_code  \
0      2016-01-03 10:30:00    2016-01-03 12.55.00    ATA     TU 32AIMN   
1      2016-01-13 15:05:00    2016-01-13 16.55.00    ATA     TU 31BIMO   
2      2016-01-16 04:10:00    2016-01-16 06.45.00    ATA     TU 32AIMN   
3      2016-01-17 14:10:00    2016-01-17 17.00.00    ATA     TU 736IOK   
4      2016-01-17 14:30:00    2016-01-17 15.50.00    ATA     TU 320IMU   

   Arrival delay  
0          260.0  
1           20.0  
2            0.0  
3            0.0  
4           22.0

In [3]:
df_airtunisia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107833 entries, 0 to 107832
Data columns (total 9 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   Filght_date               107833 non-null  object 
 1   Flight_ID                 107833 non-null  object 
 2   Departure point           107833 non-null  object 
 3   Arrival point             107833 non-null  object 
 4   Scheduled_departure_time  107833 non-null  object 
 5   Scheduled_arrival_time    107833 non-null  object 
 6   STATUS                    107833 non-null  object 
 7   Aircraft_code             107833 non-null  object 
 8   Arrival delay             107833 non-null  float64
dtypes: float64(1), object(8)
memory usage: 7.4+ MB


In [4]:
df_airtunisia.isnull().sum()

Filght_date                 0
Flight_ID                   0
Departure point             0
Arrival point               0
Scheduled_departure_time    0
Scheduled_arrival_time      0
STATUS                      0
Aircraft_code               0
Arrival delay               0
dtype: int64

In [5]:
df_airtunisia.isna().sum()

Filght_date                 0
Flight_ID                   0
Departure point             0
Arrival point               0
Scheduled_departure_time    0
Scheduled_arrival_time      0
STATUS                      0
Aircraft_code               0
Arrival delay               0
dtype: int64

In [6]:
df_airtunisia.describe()

Arrival delay
count  107833.000000
mean       48.733013
std       117.135562
min         0.000000
25%         0.000000
50%        14.000000
75%        43.000000
max      3451.000000

In [ ]:
# Use ydata-profiling to generate a report of the provided dataset
profile = ProfileReport(df_airtunisia, title='The Tunisian flight company: Tunisair dataset')
profile.to_notebook_iframe()

In [ ]:
profile.to_file("TunisAir.html")

1. skewness value of 8.43 : Many low-value observations and a few extremely high-values
2. kurtosis value (129.82) : presence of outliers, fqr from the mean (95th = 226 and max = 3451)
3. Mean = 48.73 Higher than the Median = 14, reinforcing the skewed distribution
4. Standard deviation = 117.14 / Variance (13720.74) : Spread of values, implying significant variability within the dataset.

In [27]:
# Calculer Q1 (25e percentile) et Q3 (75e percentile)
Q1 = df_airtunisia['Arrival delay'].quantile(0.25)
Q3 = df_airtunisia['Arrival delay'].quantile(0.75)
IQR = Q3 - Q1
print(f"IQR: {IQR}")

# Définir les limites
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
print(f"Lower Bound: {lower_bound}")
print(f"Upper Bound: {upper_bound}")

# Filtrer les données pour supprimer les outliers
df_cleaned = df_airtunisia[(df_airtunisia['Arrival delay'] >= lower_bound) & (df_airtunisia['Arrival delay'] <= upper_bound)]

print("Descriptive statistics for original data:")
print(df_airtunisia.describe())

print("Descriptive statistics for cleaned data:")
print(df_cleaned.describe())


IQR: 43.0
Lower Bound: -64.5
Upper Bound: 107.5
Descriptive statistics for original data:
       Arrival delay
count  107833.000000
mean       48.733013
std       117.135562
min         0.000000
25%         0.000000
50%        14.000000
75%        43.000000
max      3451.000000
Descriptive statistics for cleaned data:
       Arrival delay
count   93225.000000
mean       17.236932
std        22.331807
min         0.000000
25%         0.000000
50%         9.000000
75%        27.000000
max       107.000000


In [56]:
df_cleaned['Arrival_delay_zero'] = (df_cleaned['Arrival delay'] == 0).astype(int)
# Filter rows with zero delays
zero_delay = df_cleaned[df_cleaned['Arrival delay'] == 0]
non_zero_delay = df_cleaned[df_cleaned['Arrival delay'] != 0]

# Descriptive statistics for zero delays
print('Delay = 0')
print(zero_delay.describe())
# Descriptive statistics for non-zero delays
print('With delay')
print(non_zero_delay.describe())
print('-'*30)
print(df_cleaned.describe())

Delay = 0
       Arrival delay  Arrival_delay_zero
count        38168.0             38168.0
mean             0.0                 1.0
std              0.0                 0.0
min              0.0                 1.0
25%              0.0                 1.0
50%              0.0                 1.0
75%              0.0                 1.0
max              0.0                 1.0
With delay
       Arrival delay  Arrival_delay_zero
count   55057.000000             55057.0
mean       29.186352                 0.0
std        22.263817                 0.0
min         1.000000                 0.0
25%        12.000000                 0.0
50%        23.000000                 0.0
75%        40.000000                 0.0
max       107.000000                 0.0
------------------------------
       Arrival delay  Arrival_delay_zero
count   93225.000000        93225.000000
mean       17.236932            0.409418
std        22.331807            0.491729
min         0.000000            0.000000
25%  

C:\Users\Administrator\AppData\Local\Temp\ipykernel_2356\1792323245.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['Arrival_delay_zero'] = (df_cleaned['Arrival delay'] == 0).astype(int)


In [ ]:
non_zero_delay = non_zero_delay.drop(columns=['Arrival_delay_zero'])
profile2 = ProfileReport(non_zero_delay, title='The Tunisian flight company: cleaned dataset')
profile2.to_notebook_iframe()

In [57]:
# Détecter les mots répétés
non_zero_delay['STATUS duplicated'] = non_zero_delay['STATUS'].duplicated(keep=False)

# Convertir les mots en entiers
non_zero_delay['STATUS INT'] = pd.factorize(non_zero_delay['STATUS'])[0]
non_zero_delay = non_zero_delay.drop(columns=['STATUS'])
non_zero_delay = non_zero_delay.drop(columns=['STATUS duplicated'])
non_zero_delay.info()

<class 'pandas.core.frame.DataFrame'>
Index: 55057 entries, 1 to 107675
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Filght_date               55057 non-null  object 
 1   Flight_ID                 55057 non-null  object 
 2   Departure point           55057 non-null  object 
 3   Arrival point             55057 non-null  object 
 4   Scheduled_departure_time  55057 non-null  object 
 5   Scheduled_arrival_time    55057 non-null  object 
 6   Aircraft_code             55057 non-null  object 
 7   Arrival delay             55057 non-null  float64
 8   Arrival_delay_zero        55057 non-null  int64  
 9   STATUS INT                55057 non-null  int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 4.6+ MB


C:\Users\Administrator\AppData\Local\Temp\ipykernel_2356\3320649279.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_delay['STATUS duplicated'] = non_zero_delay['STATUS'].duplicated(keep=False)
C:\Users\Administrator\AppData\Local\Temp\ipykernel_2356\3320649279.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_zero_delay['STATUS INT'] = pd.factorize(non_zero_delay['STATUS'])[0]


In [ ]:
non_zero_delay = non_zero_delay.drop(columns=['Arrival_delay_zero'])
profile3 = ProfileReport(non_zero_delay, title='The Tunisian flight company: cleaned dataset')
profile3.to_notebook_iframe()

In [59]:
# we can drop status >>>
final_dataset = non_zero_delay.drop(columns=['STATUS INT'])
final_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 55057 entries, 1 to 107675
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Filght_date               55057 non-null  object 
 1   Flight_ID                 55057 non-null  object 
 2   Departure point           55057 non-null  object 
 3   Arrival point             55057 non-null  object 
 4   Scheduled_departure_time  55057 non-null  object 
 5   Scheduled_arrival_time    55057 non-null  object 
 6   Aircraft_code             55057 non-null  object 
 7   Arrival delay             55057 non-null  float64
dtypes: float64(1), object(7)
memory usage: 5.8+ MB


In [60]:
final_dataset.describe()

Arrival delay
count   55057.000000
mean       29.186352
std        22.263817
min         1.000000
25%        12.000000
50%        23.000000
75%        40.000000
max       107.000000

In [ ]:
profile4 = ProfileReport(final_dataset, title='The Tunisian flight company: Final dataset')
profile4.to_notebook_iframe()

Minimum: 1
Q1 (25th Percentile): 12
Median (50th Percentile): 23
Q3 (75th Percentile): 40
95th Percentile: 74
Maximum: 107
Range: 106 (difference between max and min)
Interquartile Range (IQR): 28 (difference between Q3 and Q1)
Standard Deviation: 22.26, indicating the average spread of values around the mean.
Coefficient of Variation (CV): 0.76, showing the extent of variability in relation to the mean.
Kurtosis: 1.76, indicating a relatively peaked distribution compared to a normal distribution.
Mean: 29.19, the average value.
Median Absolute Deviation (MAD): 13, indicating the average distance of data points from the median.
Skewness: 1.36, showing a right-skewed distribution.
Sum: 1,606,913, total sum of all values.
Variance: 495.68, showing the degree of spread in the data.
Monotonicity: Not monotonic, indicating the data does not consistently increase or decrease.